### PR TITLE
Add dimension table to explore

### PR DESCRIPTION
**Pull request by djbarnwal**
#883 

Some more enhancements listed here will done in a follow up PR later - #902

### CODE REVIEW

**Comment by magorlick** [comment]
Thank you for this work @djbarnwal ! I have made a few notes with their priority on UX improvements and I would like to see P0 items addressed before we ship.  Appreciate your thoughts on effort. @bcolloran , if you have any extra cycles, it would be great to get support on these items.

P0
I am missing the bar chart references that we had in the mock. I see this for your followup work for the next release, but if we could prioritize this for the current release that would be great.
image


P0
There are no indicators that let me know that I can access a detail table from the leaderboard view when there are less than 8 values.  Can we add a note in the tool tip that clicking expands the table?
image

styling:
image

copy:
"expand table"   " Click"


P0
If the user didn't name the measure, you can't tell which one it was.  Please use the expression as done in the line charts.
image
image


P0
The values are not formatted in the leaderboard as I would expect from my metrics config.
image
image


P1
Can we change this header tool tip to display the same information as you get with the line chart title hover? Can the dimension name hover display time same information as the hover over leaderboard title?

Detail table hover-over:
image

Other hover-overs:
image


P1
There is no stroke on the left side of the detail table and the column width for the value name is too wide and creates an unnecessary scroll bar interaction with a small number of measures
image


P2
I can't tell when I resort by DESC if I am seeing the bottom of the 250 available or if I am seeing the actual tail of the value metrics or what rank I am looking at if it is less than 250.  For your followup PR, we should consider adding a rank column and a tooltip for the sort icon that explains what is happening.

**Review by bcolloran — DISMISSED** [review]
Hi Dhiraj, I have a number of comments below. Overall I think this code looks good, so mostly just a few smaller things.

I see that this has ended up looking quite a bit different from the mocks; I hope we can head back in that direction with further polish passes, because I think the mocks and design spec are a bit more cohesive and polished, but I'm going to defer to @magorlick for that prioritization, so please follow her guidance on which polish/design/styling items to tackle first. I'll just mention a few things about the design

- note that default measure names are not showing up in the dimension table like they are in the linecharts. We want to make sure these match.
image 
- the summary columns should not be horizontally resizable, it's not needed for humanized numbers which will always have a narrow width (resizable colums are really only needed for long text strings), and having all the columns share the same fixed width will make it easier to scan and compare the bars when those are in.
- the table should not expand horizontally to expand the available horizontal space, it should be left-justified so that the numbers (and bars once they are ready) are easy to see next the row label without scanning across the screen. I realize all of my mocks were of a wide table with many columns, so this part of the design intent was not clear -- here is an example with only a few measure columns (note: did not remove extra measure line charts): 
image

File: src/routes/_surfaces/workspace/explore/leaderboards/DimensionLeaderboardEntrySet.svelte
Comment (djbarnwal) [reply src/routes/_surfaces/workspace/explore/leaderboards/DimensionLeaderboardEntrySet.svelte:24]: Previously we were only serving `{label, value}` array of objects throughout our codebase. Now with the addition of dimension table the datum objects are of the form `{label, measure1, measure2...}`. A generic key for storing the formatted value won't work. To define these values are internal value and to avoid conflict with measure names have adopted the double underscore variable. So formatting `{label, measure1, measure2..}` returns `{label, __formatted_measure1, __formatted_measure2, ...}`

File: src/lib/components/virtualized-table/sections/TableCells.svelte
Comment (djbarnwal) [reply src/lib/components/virtualized-table/sections/TableCells.svelte:27]: I think for this case, the code is much more readable with the current structure. Moving this to script adds unnecessary reactive complexities.

File: src/lib/components/dimension/DimensionTable.svelte
Comment (djbarnwal) [reply src/lib/components/dimension/DimensionTable.svelte:69]: Correct, only the dimension column is resizable, the measures are not.

**Comment by djbarnwal** [comment]
@magorlick Have addressed all P0 comments. Have also added row numbers to the table.

Also, the dimension column width has been designed with [these specs](https://www.notion.so/rilldata/Dimension-table-design-notes-468ba27575974072a9ed44dcbee70b7d#2aa082c6019e4eb4b43b00696d85de5f) which Hamilton and I wrote. The idea was to attain a parity with the legacy dash design where the dimension column takes rest of the container width where possible.

File: src/lib/components/dimension/DimensionTable.svelte
Comment (magorlick) [reply src/lib/components/dimension/DimensionTable.svelte:69]: If this is true, we should not be changing the cursor to the change-column-size cursor when users hover over the boundary between columns.

File: src/lib/components/virtualized-table/types.d.ts
Comment (djbarnwal) [reply src/lib/components/virtualized-table/types.d.ts:1]: These types were defined in the original Virtual Table PR so haven't touched this part of the codebase apart from restructuring the files.

File: src/lib/components/virtualized-table/sections/TableCells.svelte
Comment (bcolloran) [reply src/lib/components/virtualized-table/sections/TableCells.svelte:30]: not going to block on this, but for reference i might try something like this:

```
  const getCellProps = (row, column) => {
    const value = rows[row.index][columns[column.index]?.name];

    return {
      value,
      formattedValue:
        rows[row.index]["__formatted_" + columns[column.index]?.name],
      type: columns[column.index]?.type,
      rowActive: activeIndex === row?.index,
      suppressTooltip: scrolling,
      barValue: columns[column.index]?.total
        ? value / columns[column.index]?.total
        : 0,
      rowSelected: selectedIndex.findIndex((tgt) => row?.index === tgt) >= 0,
    };
  };
</script>

{#each virtualColumnItems as column (column.key)}
  <Row>
    {#each virtualRowItems as row (`${row.key}-${column.key}`)}
      <Cell
        {row}
        {column}
        {atLeastOneSelected}
        {...getCellProps(row, column)}
        on:inspect
        on:select-item
      />
    {/each}
  </Row>
{/each}

```

Imo that's cleaner overall because you assemble the props object in  testable code (even though we are not testing today...), and can pass it as spread props to the component. Less code overall, testability improved, and less noise in the presentational component. Not a huge deal in this case, but i think this is a good pattern and practice to shoot for

**Comment by bcolloran** [comment]
Looks good @djbarnwal. One new comment -- I'll let @magorlick confirm, but I think the idea of the row numbers is to show the sorted rank. For me, the row numbers at this point should be taken out, because they don't map to anything meaningful in the data -- since they are always shown as 1-250 in ascending order and never change, they really only say anything about the row number in the current view.

**Comment by magorlick** [comment]
> 



> Looks good @djbarnwal. One new comment -- I'll let @magorlick confirm, but I think the idea of the row numbers is to show the sorted rank. For me, the row numbers at this point should be taken out, because they don't map to anything meaningful in the data -- since they are always shown as 1-250 in ascending order and never change, they really only say anything about the row number in the current view.

Yep, the idea was that these are "leaders" or "leaderboards". So what does it mean to sort them ASC? I am seeing the tail of the leaderboards.  I wanted to understand the depth of this tail and the number of possible values to better orient in the data. We can just take this out for this release, it is a low priority.

**Comment by magorlick** [comment]
Thanks Dhiraj. These updates are looking great!  We are close.

1. I should have put together the tool tip design mock instead of describing the parts. I was thinking of the navigation tool tip styling.  This is a closer mock up for the tool tip for the leaderboard.  Could you format the tool tip to look like the other tooltips here?

image

This is the nav tooltip it would be similar to.
image



2. Could you remove row numbers? We can add rank at a later time.

**Review by nishantmonu51 — APPROVED** [review]
👍 , LGTM and all the blocking review comments seem to be addressed.

**Pull request merged**
